backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/FileSetManager.java

```diff
@@ -59,24 +67,39 @@ CompletableFuture<FileSet> save(final String prefix, final NamedFileSet files) {
 
   private CompletableFuture<FileSet.FileMetadata> saveFile(
       final String prefix, final String fileName, final Path filePath) {
+
     if (shouldCompressFile(filePath)) {
       final var algorithm = config.compressionAlgorithm().orElseThrow();
-      final var compressed = compressFile(filePath, algorithm);
-      LOG.trace("Saving compressed file {}({}) in prefix {}", fileName, compressed, prefix);
-      return client
-          .putObject(
-              put -> put.bucket(config.bucketName()).key(prefix + fileName),
-              AsyncRequestBody.fromFile(compressed))
-          .thenRunAsync(() -> cleanupCompressedFile(compressed))
-          .thenApply(unused -> FileSet.FileMetadata.withCompression(algorithm));
+      return CompletableFuture.runAsync(uploadLimit::acquireUninterruptibly)
+          .thenApply(
+              (success) -> {
+                return compressFile(filePath, algorithm);
+              })
+          .thenCompose(
+              (compressedFile) -> {
+                LOG.trace(
+                    "Saving compressed file {}({}) in prefix {}", fileName, compressedFile, prefix);
+                return client
+                    .putObject(
+                        put -> put.bucket(config.bucketName()).key(prefix + fileName),
+                        AsyncRequestBody.fromFile(compressedFile))
+                    .thenRunAsync(() -> cleanupCompressedFile(compressedFile))
+                    .thenApply(unused -> FileSet.FileMetadata.withCompression(algorithm))
+                    .whenComplete((success, error) -> uploadLimit.release());
+              });
     }
 
-    LOG.trace("Saving file {}({}) in prefix {}", fileName, filePath, prefix);
-    return client
-        .putObject(
-            put -> put.bucket(config.bucketName()).key(prefix + fileName),
-            AsyncRequestBody.fromFile(filePath))
-        .thenApply(unused -> FileSet.FileMetadata.none());
+    return CompletableFuture.runAsync(uploadLimit::acquireUninterruptibly)
+        .thenCompose(
+            (nothing) -> {
+              LOG.trace("Saving file {}({}) in prefix {}", fileName, filePath, prefix);
+              return client
+                  .putObject(
+                      put -> put.bucket(config.bucketName()).key(prefix + fileName),
+                      AsyncRequestBody.fromFile(filePath))
+                  .thenApply(unused -> FileSet.FileMetadata.none())
+                  .whenComplete((success, error) -> uploadLimit.release());
```

Comment:
:wrench: Just for symmetry, this could also be on level higher, similar to the case where we first compress the file